### PR TITLE
Aligning Parameters/ New SpectrumGenerator

### PR DESCRIPTION
**Pull request by Trapho**
## Description

The Parameters of FragmentIndex and PeptideSearchEngineAlgo do now align, such that latter can simply pass the parameters to FragmentIndex.
Added a new function to TheoreticalSpectrumGenerator, which only returns a vector of floats representing a spectrum.

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

### CODE REVIEW

File: src/openms/source/ANALYSIS/ID/FragmentIndex.cpp [hunk 1]
```diff
@@ -204,21 +204,19 @@ namespace OpenMS
...
-        for (const SimpleTSGXLMS::SimplePeak& frag : b_y_ions)
+        for (const float& frag : b_y_ions)
         {
-          if (fragment_min_mz_ > frag.mz || frag.mz > fragment_max_mz_  ) continue;
+          if (fragment_min_mz_ > frag || frag > fragment_max_mz_  ) continue;
 
-          #pragma omp critical (CreateFragment)
-          fi_fragments_.emplace_back(static_cast<UInt32>(peptide_idx),(float) frag.mz);
+         #pragma omp critical (CreateFragment)
+          fi_fragments_.emplace_back(static_cast<UInt32>(peptide_idx),(float) frag);
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/ID/FragmentIndex.cpp]:
```suggestion
          fi_fragments_.emplace_back(static_cast<UInt32>(peptide_idx), frag);
```

File: src/openms/source/ANALYSIS/ID/FragmentIndex.cpp [hunk 2]
```diff
@@ -173,7 +173,7 @@ namespace OpenMS
       tsg_params.setValue("add_x_ions", this_params.getValue("add_x_ions"));
       tsg_params.setValue("add_y_ions", this_params.getValue("add_y_ions"));
       tsg_params.setValue("add_z_ions", this_params.getValue("add_z_ions"));
-      tsg_params.setValue("add_first_prefix_ion", "true");
+      //tsg_params.setValue("add_first_prefix_ion", "true");
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/ID/FragmentIndex.cpp]:
```suggestion
```

File: src/openms/source/ANALYSIS/ID/FragmentIndex.cpp [hunk 3]
```diff
@@ -163,7 +162,8 @@ namespace OpenMS
 
       // get the spectrum generator and set the ion-types
       //TheoreticalSpectrumGenerator tsg;
-      SimpleTSGXLMS tsg;
+      //SimpleTSGXLMS tsg;
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/ID/FragmentIndex.cpp]:
```suggestion
```

File: src/openms/source/ANALYSIS/ID/FragmentIndex.cpp [hunk 4]
```diff
@@ -163,7 +162,8 @@ namespace OpenMS
 
       // get the spectrum generator and set the ion-types
       //TheoreticalSpectrumGenerator tsg;
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/ID/FragmentIndex.cpp]:
```suggestion
```

File: src/openms/include/OpenMS/CHEMISTRY/TheoreticalSpectrumGenerator.h [hunk 5]
```diff
@@ -99,6 +111,12 @@ namespace OpenMS
     /// helper to add full neutral loss ladders (for single peaks), also adds charges and ion names to the DataArrays, if the add_metainfo parameter is set to true
     void addLossesFaster_(PeakSpectrum& spectrum, double mz, const std::set<EmpiricalFormula>& f_losses, int ion_ordinal, DataArrays::StringDataArray& ion_names, DataArrays::IntegerDataArray& charges, const std::map<EmpiricalFormula, String>& formula_str_cache, double intensity, const Residue::ResidueType res_type, bool add_metainfo, int charge) const;
 
+    /// helper for getPrefixAndSuffixIonsMZ. For given Ion-Type and charge calculates a peaks
+    static void addPrefixAndSuffixIons_(std::vector< float >& spectrum, const AASequence& peptide, Residue::ResidueType res_type, int charge) ;
+
+
+
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/CHEMISTRY/TheoreticalSpectrumGenerator.h]:
```suggestion
```

File: src/openms/source/ANALYSIS/ID/FragmentIndex.cpp [hunk 6]
```diff
@@ -475,48 +483,83 @@ namespace OpenMS
     defaults_.setValue("add_z_ions", "false", "Add peaks of z-ions to the spectrum");
     defaults_.setValidStrings("add_z_ions", {"true","false"});
 
-    vector<String> all_enzymes;
-    ProteaseDB::getInstance()->getAllNames(all_enzymes);
+
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/ID/FragmentIndex.cpp]:
```suggestion
```

File: src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp [hunk 7]
```diff
@@ -17,6 +17,8 @@
 #include <OpenMS/KERNEL/MSSpectrum.h>
 #include <OpenMS/CONCEPT/RAIICleanup.h>
 
+
+
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp]:
```suggestion
```

File: src/openms/source/ANALYSIS/ID/PeptideSearchEngineAlgorithm.cpp [hunk 8]
```diff
@@ -452,12 +480,8 @@ void PeptideSearchEngineAlgorithm::postProcessHits_(const PeakMap& exp,
...
     startProgress(0, 1, "Building fragment index...");    
     FragmentIndex fragment_index_;
-    auto p = fragment_index_.getParameters();
-    p.setValue("max_processed_hits", report_top_hits_);
-    p.setValue("fragment_min_mz", param_.getValue("fragment:min_mz"));
-    p.setValue("fragment_max_mz", param_.getValue("fragment:max_mz"));
-
-    fragment_index_.setParameters(p);
+    auto this_params = getParameters();
+    fragment_index_.setParameters(this_params);
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/ID/PeptideSearchEngineAlgorithm.cpp]:
```suggestion
    auto params = getParameters();
    fragment_index_.setParameters(params);
```

File: src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp [hunk 9]
```diff
@@ -924,6 +926,105 @@ namespace OpenMS
     }
   }
 
+
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp]:
```suggestion
```

File: src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp [hunk 10]
```diff
@@ -924,6 +926,105 @@ namespace OpenMS
...
+      if (add_z_ions_)
+      {
+        addPrefixAndSuffixIons_(spectrum, peptide, Residue::ZIon, z);
+      }
+    }
+
+
+    std::sort(spectrum.begin(), spectrum.end());
+
+
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp]:
```suggestion
    std::sort(spectrum.begin(), spectrum.end());
```

File: src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp [hunk 11]
```diff
@@ -924,6 +926,105 @@ namespace OpenMS
...
+    {
+      double mono_weight(Constants::PROTON_MASS_U * charge);
+      if (peptide.hasNTerminalModification())
+      {
+        mono_weight += peptide.getNTerminalModification()->getDiffMonoMass();
+      }
+
+      switch (res_type)
+      {
+        case Residue::AIon: mono_weight += Residue::getInternalToAIon().getMonoWeight(); break;
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:979]: @jpfeuffer I vaguely recall that you changed something in the TSG regarding the calculation of internal masses. Does this code also need this update?

File: src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp [hunk 12]
```diff
@@ -924,6 +926,105 @@ namespace OpenMS
...
+        default: break;
+      }
+
+      Size i = 0;
+      for (; i < peptide.size(); ++i)
+      {
+        mono_weight += peptide[i].getMonoWeight(Residue::Internal);
+        double pos(mono_weight / charge);
+        spectrum.emplace_back(pos);
+
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp]:
```suggestion
```

File: src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp [hunk 13]
```diff
@@ -924,6 +926,105 @@ namespace OpenMS
...
+        case Residue::BIon: mono_weight += Residue::getInternalToBIon().getMonoWeight(); break;
+        case Residue::CIon: mono_weight += Residue::getInternalToCIon().getMonoWeight(); break;
+        default: break;
+      }
+
+      Size i = 0;
+      for (; i < peptide.size(); ++i)
+      {
+        mono_weight += peptide[i].getMonoWeight(Residue::Internal);
+        double pos(mono_weight / charge);
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:995]:
```suggestion
        float pos(mono_weight / charge);
```

File: src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp [hunk 14]
```diff
@@ -924,6 +926,105 @@ namespace OpenMS
...
+        spectrum.emplace_back(pos);
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp]:
```suggestion
        spectrum.emplace_back(mono_weight / charge);
```

File: src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp [hunk 15]
```diff
@@ -924,6 +926,105 @@ namespace OpenMS
...
+        case Residue::ZIon: mono_weight += Residue::getInternalToZIon().getMonoWeight(); break;
+        default: break;
+      }
+
+      for (Size j = peptide.size(); j >= 1; --j)
+      {
+        mono_weight += peptide[j-1].getMonoWeight(Residue::Internal);
+        double pos(mono_weight / charge);
+
+        spectrum.emplace_back(pos);
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp]:
```suggestion
        spectrum.emplace_back(mono_weight / charge);
```

File: src/tests/class_tests/openms/source/FragmentIndex_test.cpp [hunk 16]
```diff
@@ -341,17 +340,36 @@ START_SECTION(tolerance)
...
 
+tolTest.build(entries);
+
+TheoreticalSpectrumGenerator tsg2;
+
+
+
+
+
+
```
Comment (timosachsenberg) [inline src/tests/class_tests/openms/source/FragmentIndex_test.cpp]:
```suggestion
```

File: src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp [hunk 17]
```diff
@@ -924,6 +924,97 @@ namespace OpenMS
...
+        case Residue::YIon: mono_weight += Residue::getInternalToYIon().getMonoWeight(); break;
+        case Residue::ZIon: mono_weight += Residue::getInternalToZIon().getMonoWeight(); break;
+        default: break;
+      }
+
+      for (Size j = peptide.size(); j >= 1; --j)
+      {
+        mono_weight += peptide[j-1].getMonoWeight(Residue::Internal);
+        spectrum.emplace_back(mono_weight / charge);
+
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp]:
```suggestion
        spectrum.emplace_back((float)mono_weight / charge);
```

File: src/openms/source/ANALYSIS/ID/FragmentIndex.cpp [hunk 18]
```diff
@@ -455,6 +457,9 @@ namespace OpenMS
       trimHits(sms);
   }
 
+
+
+
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/ID/FragmentIndex.cpp]:
```suggestion
```

File: src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp (hunk 11 above)
Comment (jpfeuffer) [reply src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:979]: I think it was not merged yet

Comment (timosachsenberg) [reply src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:979]: Let em check... I think https://github.com/OpenMS/OpenMS/pull/7226/files was merged through https://github.com/OpenMS/OpenMS/pull/7237/files just with some additional fixes. Was there another one?

Comment (timosachsenberg) [reply src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:979]: ah it was here. https://github.com/OpenMS/OpenMS/pull/7237/files#diff-11be5715633a0e04aa066f79aa957d602ed73dc63e9d9fb2239d270356ede638R414 so it already got merged... all good

**Pull request merged**